Add apps/v1beta1 deployments with new defaults

soltysh opened the pull request:
This pull introduces deployments under `apps/v1beta1` and fixes #23597 and #23304.

TODO:

* [x] - create new type `apps/v1beta1.Deployment`
* [x] - update kubectl (stop, scale)
* [ ] - ~~new `kubectl run` generator~~ - this will only duplicate half of generator code, I suggest replacing current to use new endpoint
* [ ] - ~~create extended tests~~ - I've added integration and cmd tests verifying new endpoints
* [ ] - ~~create `hack/test-update-storage-objects.sh`~~ - see above

This is currently blocked by https://github.com/kubernetes/kubernetes/pull/38071, due to conflicting name `v1beta1.Deployment`. 

```release-note
Introduce apps/v1beta1.Deployments resource with modified defaults compared to extensions/v1beta1.Deployments.
```

@kargakis @mfojtik @kubernetes/sig-apps-misc

soltysh committed:
Deployments under apps/v1beta1 with new defaults

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39683)
<!-- Reviewable:end -->

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -138,7 +138,7 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule("impersonate").Groups(legacyGroup).Resources("serviceaccounts").RuleOrDie(),
 
-				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets").RuleOrDie(),
+				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets", "deployments", "deployments/scale").RuleOrDie(),
```
put the deployment objects in their own rule

liggitt replied:
actually, nevermind, misread the later patterns

kargakis commented on pkg/apis/apps/v1beta1/defaults.go:
```diff
@@ -45,3 +47,47 @@ func SetDefaults_StatefulSet(obj *StatefulSet) {
+	if obj.Spec.RevisionHistoryLimit == nil {
+		obj.Spec.RevisionHistoryLimit = new(int32)
+		*obj.Spec.RevisionHistoryLimit = 2
+	}
```
Let's add a default for progressDeadlineSeconds too since the Progressing condition seems to be useful and can be used in different contexts. In OpenShift, we use a default of 10 minutes.

kargakis replied:
@kubernetes/api-reviewers

kargakis replied:
I am planning to write up an external controller that utilizes the Progressing condition produced by ProgressDeadlineSeconds and kubectl rollout undo to showcase autorollback.

soltysh commented:
@kargakis @kubernetes/sig-apps-pr-reviews this is ready for review

soltysh commented on pkg/apis/apps/v1beta1/types.go:
```diff
@@ -106,3 +107,224 @@ type StatefulSetList struct {
+// +genclient=true
+
+// Deployment enables declarative updates for Pods and ReplicaSets.
+type Deployment struct {
```
@kubernetes/api-reviewers this and defaults needs your approval

janetkuo replied:
You need to update comments wrt the new defaults

soltysh commented:
I've updated initial comments, but I'll also leave an info for reviewer.

1. I didn't want to introduce another generator, specifically for `apps/v1beta1.Deployments` because that will duplicate majority of methods in the generator. I'd suggest updating the current generator to use the new endpoint, @kubernetes/sig-cli-pr-reviews wdyt?
2. I didn't add any e2e tests nor storage upgrade, b/c I went with an integration tests for verifying if the object is reachable from both endpoints.

soltysh commented:
Weird error, I've rebased against latest master and waiting for test results.

kargakis commented on pkg/apis/apps/v1beta1/conversion.go:
```diff
@@ -112,3 +140,156 @@ func Convert_apps_StatefulSetSpec_To_v1beta1_StatefulSetSpec(in *apps.StatefulSe
+}
+
+func Convert_v1beta1_RollingUpdateDeployment_To_extensions_RollingUpdateDeployment(in *RollingUpdateDeployment, out *extensions.RollingUpdateDeployment, s conversion.Scope) error {
+	if err := s.Convert(in.MaxUnavailable, &out.MaxUnavailable, 0); err != nil {
```
Is this the reason behind the custom conversions? Needs a comment https://github.com/kubernetes/kubernetes/issues/39865

soltysh replied:
Discussed this with Stefan and Michalis, will double check if these are needed.

soltysh replied:
It looks like this is not working without it, I got error complaining about not having a conversion between `intstr.IntOrString` and `*intstr.IntOrString`. I'll add a TODO here.

kargakis replied:
@kubernetes/sig-api-machinery-misc any reason we can't have a conversion func for intstr.IntOrString->*intstr.IntOrString?

soltysh replied:
I've moved that discussion over to #39865.

kargakis commented on pkg/apis/apps/v1beta1/defaults.go:
```diff
@@ -45,3 +47,51 @@ func SetDefaults_StatefulSet(obj *StatefulSet) {
+	if strategy.Type == "" {
+		strategy.Type = RollingUpdateDeploymentStrategyType
+	}
+	if strategy.Type == RollingUpdateDeploymentStrategyType || strategy.RollingUpdate != nil {
```
Should we set defaults even when the type is Recreate?

soltysh replied:
Discussed with Michalis on IRC, we'll check just the Type.

kargakis commented on pkg/apis/apps/v1beta1/register.go:
```diff
@@ -41,6 +41,10 @@ var (
+		&Deployment{},
+		&DeploymentList{},
+		&DeploymentRollback{},
+		&Scale{},
```
It is really unfortunate that we need to drag this in every place.

soltysh replied:
Yeah, but there's a separate issue to solve it, can't recall from the top of my head.

kargakis replied:
Didn't mean to solve it here, just a general comment

kargakis commented on pkg/apis/apps/v1beta1/types.go:
```diff
@@ -106,3 +152,224 @@ type StatefulSetList struct {
+	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty" protobuf:"varint,6,opt,name=revisionHistoryLimit"`
+
+	// Indicates that the deployment is paused and will not be processed by the
+	// deployment controller.
```
This comment is no longer true, mind updating it? Paused deployments are still processed by the deployment controller, we just don't perform upgrades.

soltysh replied:
Updated.

kargakis commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -123,7 +123,7 @@ func ClusterRoles() []rbac.ClusterRole {
 				rbac.NewRule("impersonate").Groups(legacyGroup).Resources("serviceaccounts").RuleOrDie(),
 
-				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets").RuleOrDie(),
+				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets", "deployments", "deployments/scale").RuleOrDie(),
```
Is this missing rollback and status subresources?

kargakis replied:
Same comment for the rest of the roles - except `view` does not need rollback.

soltysh replied:
I haven't seen status being added to any of the resources, I'll add rollback to admin and edit, though.

kargakis replied:
Yes, status is not needed in these roles

kargakis commented on test/integration/master/master_test.go:
```diff
@@ -195,6 +253,59 @@ func TestAutoscalingGroupBackwardCompatibility(t *testing.T) {
+		{"POST", extensionsPath("deployments", metav1.NamespaceDefault, ""), deploymentExtensions, integration.Code201, ""},
+		{"GET", extensionsPath("deployments", metav1.NamespaceDefault, "test-deployment"), "", integration.Code200, testapi.Extensions.GroupVersion().String()},
+		{"GET", appsPath("deployments", metav1.NamespaceDefault, "test-deployment"), "", integration.Code200, testapi.Apps.GroupVersion().String()},
+		{"DELETE", extensionsPath("deployments", metav1.NamespaceDefault, "test-deployment"), "", integration.Code200, api.Registry.GroupOrDie(v1.GroupName).GroupVersion.String()},
```
Can you test GETs from both endpoints after DELETE?

soltysh replied:
I don't think this is needed. The 2nd POST is happening after DELETE, and since both endpoints share the same storage it should fail POST if the previous deletion failed, assuming I'm using the same name for both deployments, and as you can see, I am..

soltysh replied:
I did more tests, and the GET after DELETE will actually return you the object is being still deleted. That's why I'm also forced to use different names for both.

soltysh commented:
Comments addressed and rebased.

@kargakis @janetkuo ptal

kargakis commented:
@kubernetes/sig-apps-pr-reviews @kubernetes/sig-api-machinery-pr-reviews it would be great to have this for 1.6 as the new defaults would help a lot with existing Deployment issues. Any reviews to move this forward would be appreciated.

soltysh commented:
Rebased.

soltysh commented:
Updated the policy tests.

smarterclayton commented on pkg/apis/apps/v1beta1/defaults.go:
```diff
@@ -45,3 +47,51 @@ func SetDefaults_StatefulSet(obj *StatefulSet) {
 	}
 }
+
+func SetDefaults_Deployment(obj *Deployment) {
```
Add comments about how this differs from v1beta1 - call out how these differ.

smarterclayton commented on pkg/apis/apps/v1beta1/types.go:
```diff
@@ -26,6 +27,51 @@ const (
 	StatefulSetInitAnnotation = "pod.alpha.kubernetes.io/initialized"
 )
 
+// describes the attributes of a scale subresource
```
Doesn't match godoc form.

liggitt commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -216,6 +218,8 @@ func Run(s *options.ServerRunOptions) error {
 		return fmt.Errorf("error in initializing storage factory: %s", err)
 	}
+	// keep Deployments in extensions for backwards compatibility, we'll have to migrate at some point, eventually
+	storageFactory.AddCohabitatingResources(extensions.Resource("deployments"), apps.Resource("deployments"))
```
No, otherwise existing deployments will disappear on upgrade

soltysh replied:
I've put that information in the comment above.

soltysh commented:
Rebased and addressed Clayton's comments.

soltysh commented:
Fixed bazel build.

kargakis commented on pkg/apis/apps/v1beta1/types_swagger_doc_generated.go:
```diff
@@ -27,6 +27,143 @@ package v1beta1
+	"revisionHistoryLimit":    "The number of old ReplicaSets to retain to allow rollback. This is a pointer to distinguish between explicit zero and not specified.",
+	"paused":                  "Indicates that the deployment is paused.",
+	"rollbackTo":              "The config this deployment is rolling back to. Will be cleared after rollback is done.",
+	"progressDeadlineSeconds": "The maximum time in seconds for a deployment to make progress before it is considered to be failed. The deployment controller will continue to process failed deployments and a condition with a ProgressDeadlineExceeded reason will be surfaced in the deployment status. Once autoRollback is implemented, the deployment controller will automatically rollback failed deploym ...
```
Now it is set by default so the comment needs to be updated.

soltysh replied:
Yeah, I'll also update the other fields where defaults have changed.

kargakis commented on pkg/apis/apps/v1beta1/zz_generated.defaults.go:
```diff
@@ -29,11 +29,151 @@ import (
+func SetObjectDefaults_Deployment(in *Deployment) {
+	SetDefaults_Deployment(in)
+	v1.SetDefaults_PodSpec(&in.Spec.Template.Spec)
+	for i := range in.Spec.Template.Spec.Volumes {
```
Why do we need to do this?

soltysh replied:
That's a generated code.

kargakis commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -123,14 +123,16 @@ func ClusterRoles() []rbac.ClusterRole {
 
-				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets").RuleOrDie(),
+				rbac.NewRule(ReadWrite...).Groups(appsGroup).Resources("statefulsets",
+					"deployments", "deployments/scale", "deployments/rollback").RuleOrDie(),
```
So the admin seems to be able to update status for other resources (eg. replicationcontrollers)

kargakis replied:
Sorry, only read

kargakis replied:
edit and view also need to be able to read /status

soltysh replied:
Double-checked with @deads2k and it's not necessary. Since they can get that information by getting the entire object. If it's needed we can always add afterwards.

kargakis commented:
@soltysh you have picked up a bunch of cruft in staging that I am not sure is needed

soltysh commented:
> @soltysh you have picked up a bunch of cruft in staging that I am not sure is needed

That's a result of running `hack/update-staging-client.sh` @sttts is currently working on making this check obligatory, without it our staging client is waaay out of sync with what it should be. That can be seen in that last commit.

soltysh commented:
This is currently waiting for https://github.com/kubernetes/kubernetes/pull/42068 to merge. I'll rebase on top of it.

janetkuo commented on pkg/apis/apps/v1beta1/types.go:
```diff
@@ -106,3 +152,223 @@ type StatefulSetList struct {
+	// that the total number of pods available at all times during the update is at
+	// least 70% of desired pods.
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,1,opt,name=maxUnavailable"`
```
You need to update comments wrt the new defaults

soltysh replied:
Yeah, I have that done locally. I'm currently waiting for https://github.com/kubernetes/kubernetes/pull/42068 to merge and will re-generate everything and push updated version.

soltysh committed:
Generated changes for apps/v1beta1.deployments

soltysh committed:
Staging changes for apps/v1beta1.deployments

soltysh commented:
Comments addressed and reabsed. @janetkuo mind taking another look?

janetkuo commented:
Generally LGTM, will wait for https://github.com/kubernetes/kubernetes/pull/39683#discussion_r103085178 being addressed before applying LGTM tag. 

New defaults need approval from @bgrant0607 or other @kubernetes/api-reviewers 

Also note that `kubectl rollout status`, `history`, and `undo` needs to be updated to support deployment.apps, but can be a follow up PR.

janetkuo commented:
/lgtm

I saw that https://github.com/kubernetes/kubernetes/pull/39683#discussion_r103085178 is addressed

soltysh commented:
Rebased, will re-apply lgtm label when tests are green.

soltysh commented:
@k8s-bot bazel test this

soltysh committed:
Updated linted packages

soltysh commented:
Updated `hack/.linted_packages`.

soltysh commented:
@k8s-bot unit test this

bgrant0607 commented:
Defaults and API group LGTM.

/approve

janetkuo commented:
@k8s-bot unit test this

janetkuo commented:
/lgtm

soltysh commented:
@janetkuo there seems to be some problem with test-cmd, I'm currently looking into it.

janetkuo commented:
@soltysh also note that `kubectl run` with `--generator=deployment/v1beta1`now generates apps/v1beta1 deployments, with new defaults. `kubectl run` should instead provide the same behavior if the users pin to a specific generator version.

janetkuo commented:
> there seems to be some problem with test-cmd, I'm currently looking into it.

@soltysh it's not you. AWS s3 went down today. 

```
W0228 10:57:11.447] + ./hack/install-etcd.sh
W0228 10:59:19.387] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 10:59:19.488] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 4 retries left.
W0228 11:01:27.771] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 11:01:27.872] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 3 retries left.
W0228 11:03:36.283] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 11:03:36.384] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 2 retries left.
W0228 11:05:44.795] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 11:05:44.896] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 1 retries left.
W0228 11:07:53.179] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 11:07:53.280] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 0 retries left.
```

janetkuo commented:
@k8s-bot unit test this

kargakis commented:
again needs a rebase :/

soltysh commented:
> @soltysh also note that kubectl run with --generator=deployment/v1beta1now generates apps/v1beta1 deployments, with new defaults. kubectl run should instead provide the same behavior if the users pin to a specific generator version.

I initially didn't want to create a new generator due to quite extensive copy&paste (which we get rid of just recently when removing extensions/v1beta1 jobs) but I guess there's no other option. The only problem I have is the name :/ I'll open a PR later today to address that and we can discuss more then.

soltysh commented:
Rebased, on green tests I'll apply lgtm label and bump the prio up, I've paid my due doing rebase at least 3 or 4 times already, including 2 this morning :/

soltysh commented:
> @soltysh it's not you. AWS s3 went down today.

Nope, it looks like it's still test-cmd.

soltysh commented:
> Also note that kubectl rollout status, history, and undo need to be updated to support deployment.apps, but can be a follow up PR.

Actually that's guarded in our test-cmd and that was the reason they were failing, additionally `autoscale`. I'm adding those and pushing changes.

soltysh commented:
Rebased and added the changes I've mentioned in the previous comment. On green tests I'm re-applying lgtm label and will bump prio to p2, to have it merged.

soltysh commented:
@k8s-bot non-cri e2e test this

ethernetdan commented:
@soltysh an exception was submitted for this PR and had yet to be approved, is there a reason this was merged?

@kubernetes/kubernetes-release-managers

soltysh commented:
@ethernetdan I apologize I understood this was already approved.

calebamiles commented:
PR to revert the merge: https://github.com/kubernetes/kubernetes/pull/42368
